src/main/scala/org/broadinstitute/dsde/workbench/sam/model/SamModel.scala

```diff
@@ -141,6 +141,29 @@ object UserStatusDetails {
   // Ideally we'd just store this boolean in a lazy val, but this will upset the spray/akka json serializers
   // I can't imagine a scenario where we have enough action patterns that would make this def discernibly slow though
   def isAuthDomainConstrainable: Boolean = actionPatterns.exists(_.authDomainConstrainable)
+
+  /** Recursively get all actions for a role, including actions from included roles
+    * @param roleName
+    *   role to get actions for
+    * @param visitedRoles
+    *   set of roles that have already been visited to prevent infinite recursion
+    * @return
+    *   set of all actions for the role
+    */
+  def getRoleActions(roleName: ResourceRoleName, visitedRoles: Set[ResourceRoleName] = Set.empty): Set[ResourceAction] =
+    if (visitedRoles.contains(roleName)) {
+      // prevent infinite recursion
+      Set.empty
+    } else {
+      roles
+        .find(_.roleName == roleName)
+        .map(role =>
+          role.includedRoles.foldLeft(role.actions)((accumulatedActions, includedRole) =>
+            accumulatedActions ++ getRoleActions(includedRole, visitedRoles + roleName)
+          )
+        )
+        .getOrElse(Set.empty)
+    }
```

Comment:
I _think_ there's a corner case here where the recursion exits early:

Role "one" includes roles "two" and "three"

Role "two" includes roles "A", "B", "C"

Role "three" includes roles "B", "C", "D", "E"

As soon as the recursion hits "B" for the second time, it will exit - potentially never considering "D" or "E" (yeah, it's a Set, it won't process alphabetically, but hopefully this gets the point across)

I also wonder if this corner case is even possible in the current `reference.conf` and how important it is to guard against

Reply:
I am not following why there would be an early exit. The `foldLeft` will evaluate all `includedRoles`. Roles "two" and "three" evaluate in separate call trees so in your example "B" is actually visited twice. `visitedRoles` does not prevent multiple visits, just cycles.

Maybe you are thinking that `visitedRoles` has a broader scope than it does. Each call to `getRoleActions` gets its own copy of `visitedRoles`. `visitedRoles + roleName` does not modify `visitedRoles` but creates a new Set with `roleName` added.

Reply:
yup, I see this now, I was misreading the recursion, all good!